### PR TITLE
Core/Warden: More Warden improvements

### PR DESCRIPTION
**Pull request by Treeston**
* Make `LUA_STR_CHECK` functional and (somewhat more) useful by turning it into `LUA_EVAL_CHECK`
* Re-structure and cleanup how we handle Warden scheduling
* Three new checks, one of which should catch naive Lua unlockers, two of which should catch PQR (ref #23035)

_(Special thanks to @scizzydo for Warden insights.)_

### CODE REVIEW

**Comment by Warpten** [comment]
No.

File: sql/updates/world/3.3.5/9999_99_99_99_world_WARDEN.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+-- (left join is for reapply safety)
+UPDATE `warden_checks` wc1
+    LEFT JOIN `warden_checks` wc2 ON `wc2`.`id` = 0
+    SET `wc1`.`id` = `wc1`.`id`-1
+    WHERE `wc2`.`id` IS NULL;
+
+DELETE FROM `warden_checks` WHERE `id` IN (786,787,788);
+INSERT INTO `warden_checks` (`id`,`type`,`str`,`address`,`length`,`result`, `comment`) VALUES
+(786, 139, 'forceinsecure() return issecure()', NULL, NULL, NULL, 'Detects naive Lua unlockers'),
+(787, 243, NULL, CONV('819210',16,10), 16, UNHEX('558BEC518305A013D40001A19C13D400'), 'FrameScript::Execute'),
```
Comment (Shauren) [inline sql/updates/world/3.3.5/9999_99_99_99_world_WARDEN.sql:16]: you know you can just write 0x819210 and mysql will insert it properly?

File: src/server/game/Warden/WardenWin.cpp [hunk 2]
```diff
@@ -413,26 +421,11 @@ void WardenWin::HandleCheckResult(ByteBuffer &buff)
                 TC_LOG_DEBUG("warden", "RESULT %s passed CheckId %u account Id %u", EnumUtils::ToConstant(check.Type), id, _session->GetAccountId());
                 break;
             }
-            case LUA_STR_CHECK:
+            case LUA_EVAL_CHECK:
             {
-                uint8 Lua_Result;
-                buff >> Lua_Result;
+                buff.read<uint8>();
```
Comment (Warpten) [inline src/server/game/Warden/WardenWin.cpp:426]: you need to read_skip the rest if != 0

File: src/server/game/Warden/WardenCheckMgr.h [hunk 3]
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
...
-    PAGE_CHECK_A = 0xB2, // 178: uint Seed + byte[20] SHA1 + uint Addr + byte Len (scans all pages for specified hash)
-    PAGE_CHECK_B = 0xBF, // 191: uint Seed + byte[20] SHA1 + uint Addr + byte Len (scans only pages starts with MZ+PE headers for specified hash)
-    MPQ_CHECK = 0x98, // 152: byte fileNameIndex (check to ensure MPQ file isn't modified)
-    LUA_STR_CHECK = 0x8B, // 139: byte luaNameIndex (check to ensure LUA string isn't used)
-    DRIVER_CHECK = 0x71, // 113: uint Seed + byte[20] SHA1 + byte driverNameIndex (check to ensure driver isn't loaded)
-    TIMING_CHECK = 0x57, //  87: empty (check to ensure GetTickCount() isn't detoured)
-    PROC_CHECK = 0x7E, // 126: uint Seed + byte[20] SHA1 + byte moluleNameIndex + byte procNameIndex + uint Offset + byte Len (check to ensure proc isn't detoured)
-    MODULE_CHECK = 0xD9  // 217: uint Seed + byte[20] SHA1 (check to ensure module isn't injected)
+    NONE_CHECK     =   0, // SKIP
+    TIMING_CHECK   =  87, // nyi
```
Comment (Warpten) [inline src/server/game/Warden/WardenCheckMgr.h:49]: why nyi, it's always sent (don't ask me why). it's an empty payload for the request

File: src/server/game/Warden/WardenCheckMgr.h [hunk 4]
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
...
+        case MPQ_CHECK:      return MODDED_CHECK_CATEGORY;
+        case PAGE_CHECK_A:   return INJECT_CHECK_CATEGORY;
+        case PAGE_CHECK_B:   return INJECT_CHECK_CATEGORY;
+        case MODULE_CHECK:   return INJECT_CHECK_CATEGORY;
+        case MEM_CHECK:      return MODDED_CHECK_CATEGORY;
+        default:             return NUM_CHECK_CATEGORIES;
+    }
+}
+
+inline constexpr WorldIntConfigs GetWardenCategoryCountConfig(WardenCheckCategory category)
```
Comment (Warpten) [inline src/server/game/Warden/WardenCheckMgr.h:77]: since c++17 inline is implicitely specified

File: src/server/game/Warden/WardenCheckMgr.h [hunk 5]
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
...
+    PROC_CHECK     = 126, // nyi
+    LUA_EVAL_CHECK = 139, // evaluate arbitrary Lua check
+    MPQ_CHECK      = 152, // get hash of MPQ file (to check it is not modified)
+    PAGE_CHECK_A   = 178, // scans all pages for specified SHA1 hash
+    PAGE_CHECK_B   = 191, // scans only pages starts with MZ+PE headers for specified hash
+    MODULE_CHECK   = 217, // check to make sure module isn't injected
+    MEM_CHECK      = 243, // retrieve specific memory
 };
 
+inline constexpr WardenCheckCategory GetWardenCheckCategory(WardenCheckType type)
```
Comment (Warpten) [inline src/server/game/Warden/WardenCheckMgr.h:60]: same as L77

File: src/server/game/Warden/WardenCheckMgr.h (hunk 3 above)
Comment (Treeston) [reply src/server/game/Warden/WardenCheckMgr.h:49]: the result is not processed

File: src/server/game/Warden/WardenWin.cpp (hunk 2 above)
Comment (Treeston) [reply src/server/game/Warden/WardenWin.cpp:426]: No, in fact current Warden is broken, and adding a `LUA_STR_CHECK` to the DB in current Warden will corrupt data after it.

Result should only be read if `Lua_Result` is zero, not nonzero.

**Comment by Treeston** [comment]
Split off a number of unrelated refactors into 5b4288f, a3971ca, fc32c3f.

Rebased this PR onto the result. This gives us a significantly more manageable changeset to review.

File: src/server/game/Warden/WardenCheckMgr.h (hunk 3 above)
Comment (Warpten) [reply src/server/game/Warden/WardenCheckMgr.h:49]: it is (or was)

File: src/server/game/Warden/WardenWin.cpp (hunk 2 above)
Comment (Warpten) [reply src/server/game/Warden/WardenWin.cpp]: i mean yeah i know the check is broken as far as its implementation in module is concerned but then you still have data if = 0

Comment (Treeston) [reply src/server/game/Warden/WardenWin.cpp]: I have only ever observed result 1 for `::Execute` substitution, I believe it to be the return value from one of the lua library functions.

Anyway, added `read_skip`.

File: src/server/game/Warden/WardenCheckMgr.h (hunk 3 above)
Comment (Treeston) [reply src/server/game/Warden/WardenCheckMgr.h:49]: The actual timing result is never used, just logged.

File: sql/updates/world/3.3.5/9999_99_99_99_world_WARDEN.sql (hunk 1 above)
Comment (Warpten) [reply sql/updates/world/3.3.5/9999_99_99_99_world_WARDEN.sql]: for binary i personally do `x'CAFEBABE'` and it just works

**Pull request merged**